src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,209 @@
+        }
+        Node<T> nodeWithInputIndex = getNodeNodeByIndex(index);
+        return nodeWithInputIndex.getItem();
+    }
+
+    private Node<T> getNodeNodeByIndex(int index) {
+        if (size / 2 > index) {
+            return getNodeFromFirstHalfOfList(index);
+        } else {
+            return getNodeFromSecondHalfOfList(index);
+        }
+    }
+
+    private Node<T> getNodeFromFirstHalfOfList(int index) {
+        int indexCounter = 0;
+        Node<T> nodeWithInputIndex = first;
+        while (indexCounter < index) {
+            nodeWithInputIndex = nodeWithInputIndex.getNext();
+            indexCounter++;
+        }
+        return nodeWithInputIndex;
+    }
+
+    private Node<T> getNodeFromSecondHalfOfList(int index) {
+        int indexCounter = size - 1;
+        Node<T> nodeWithInputIndex = last;
+        while (indexCounter > index) {
+            nodeWithInputIndex = nodeWithInputIndex.getPrev();
+            indexCounter--;
+        }
+        return nodeWithInputIndex;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(MESSAGE_INDEX_OUT_OF_BOUNDS_EXCEPTION);
+        }
+        Node<T> actualNode = getNodeNodeByIndex(index);
+        T oldValue = actualNode.getItem();
+        actualNode.setItem(value);
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(MESSAGE_INDEX_OUT_OF_BOUNDS_EXCEPTION);
+        }
+        Node<T> nodeForDeletion = getNodeNodeByIndex(index);
+        unlinkNode(nodeForDeletion);
+        return nodeForDeletion.getItem();
     }
 
     @Override
-    public boolean remove(T object) {
-        return false;
+    public boolean remove(T item) {
+        Node<T> nodeForDeletion = getNodeByItem(item);
+        if (nodeForDeletion == null) {
+            return false;
+        }
+        unlinkNode(nodeForDeletion);
+        return true;
+    }
+
+    private void unlinkNode(Node<T> nodeForDeletion) {
+        if (nodeForDeletion == first) {
+            deleteFirstNode(nodeForDeletion);
+        } else if (nodeForDeletion == last) {
+            deleteLastNode(nodeForDeletion);
+        } else {
+            nodeForDeletion.getNext().setPrev(nodeForDeletion.getPrev());
+            nodeForDeletion.getPrev().setNext(nodeForDeletion.getNext());
+        }
+        size--;
+    }
+
+    private void deleteLastNode(Node<T> nodeForDeletion) {
+        nodeForDeletion.getPrev().setNext(null);
+        last = nodeForDeletion.getPrev();
+    }
+
+    private void deleteFirstNode(Node<T> nodeForDeletion) {
+        first = nodeForDeletion.getNext();
+        if (first != null) {
+            nodeForDeletion.getNext().setPrev(null);
+        }
+    }
+
+    private Node<T> getNodeByItem(T item) {
+        Node<T> nodeForDeletion = first;
```

Comment:
```suggestion
        Node<T> node = first;
```
in this case you should name more universally, maybe you'll need this method not only for deletion